commit fa4983862e506d395acc1b8d14dbebf63acc2e82
Author: Weston Andros Adamson <dros@primarydata.com>
Date:   Thu Sep 5 15:09:50 2024 -0400

    nfsd: add LOCALIO support
    
    Add server support for bypassing NFS for localhost reads, writes, and
    commits. This is only useful when both the client and server are
    running on the same host.
    
    If nfsd_open_local_fh() fails then the NFS client will both retry and
    fallback to normal network-based read, write and commit operations if
    localio is no longer supported.
    
    Care is taken to ensure the same NFS security mechanisms are used
    (authentication, etc) regardless of whether localio or regular NFS
    access is used.  The auth_domain established as part of the traditional
    NFS client access to the NFS server is also used for localio.  Store
    auth_domain for localio in nfsd_uuid_t and transfer it to the client
    if it is local to the server.
    
    Relative to containers, localio gives the client access to the network
    namespace the server has.  This is required to allow the client to
    access the server's per-namespace nfsd_net struct.
    
    This commit also introduces the use of NFSD's percpu_ref to interlock
    nfsd_destroy_serv and nfsd_open_local_fh, to ensure nn->nfsd_serv is
    not destroyed while in use by nfsd_open_local_fh and other LOCALIO
    client code.
    
    CONFIG_NFS_LOCALIO enables NFS server support for LOCALIO.
    
    Signed-off-by: Weston Andros Adamson <dros@primarydata.com>
    Signed-off-by: Trond Myklebust <trond.myklebust@hammerspace.com>
    Co-developed-by: Mike Snitzer <snitzer@kernel.org>
    Signed-off-by: Mike Snitzer <snitzer@kernel.org>
    Co-developed-by: NeilBrown <neilb@suse.de>
    Signed-off-by: NeilBrown <neilb@suse.de>
    Reviewed-by: Jeff Layton <jlayton@kernel.org>
    Acked-by: Chuck Lever <chuck.lever@oracle.com>
    Signed-off-by: Anna Schumaker <anna.schumaker@oracle.com>

diff --git a/fs/nfsd/nfsctl.c b/fs/nfsd/nfsctl.c
index 64c1b4d649bc..3adbc05ebaac 100644
--- a/fs/nfsd/nfsctl.c
+++ b/fs/nfsd/nfsctl.c
@@ -18,6 +18,7 @@
 #include <linux/sunrpc/svc.h>
 #include <linux/module.h>
 #include <linux/fsnotify.h>
+#include <linux/nfslocalio.h>
 
 #include "idmap.h"
 #include "nfsd.h"
@@ -2257,7 +2258,9 @@ static __net_init int nfsd_net_init(struct net *net)
 	get_random_bytes(&nn->siphash_key, sizeof(nn->siphash_key));
 	seqlock_init(&nn->writeverf_lock);
 	nfsd_proc_stat_init(net);
-
+#if IS_ENABLED(CONFIG_NFS_LOCALIO)
+	INIT_LIST_HEAD(&nn->local_clients);
+#endif
 	return 0;
 
 out_repcache_error:
@@ -2268,6 +2271,22 @@ static __net_init int nfsd_net_init(struct net *net)
 	return retval;
 }
 
+#if IS_ENABLED(CONFIG_NFS_LOCALIO)
+/**
+ * nfsd_net_pre_exit - Disconnect localio clients from net namespace
+ * @net: a network namespace that is about to be destroyed
+ *
+ * This invalidated ->net pointers held by localio clients
+ * while they can still safely access nn->counter.
+ */
+static __net_exit void nfsd_net_pre_exit(struct net *net)
+{
+	struct nfsd_net *nn = net_generic(net, nfsd_net_id);
+
+	nfs_uuid_invalidate_clients(&nn->local_clients);
+}
+#endif
+
 /**
  * nfsd_net_exit - Release the nfsd_net portion of a net namespace
  * @net: a network namespace that is about to be destroyed
@@ -2285,6 +2304,9 @@ static __net_exit void nfsd_net_exit(struct net *net)
 
 static struct pernet_operations nfsd_net_ops = {
 	.init = nfsd_net_init,
+#if IS_ENABLED(CONFIG_NFS_LOCALIO)
+	.pre_exit = nfsd_net_pre_exit,
+#endif
 	.exit = nfsd_net_exit,
 	.id   = &nfsd_net_id,
 	.size = sizeof(struct nfsd_net),
@@ -2322,6 +2344,7 @@ static int __init init_nfsd(void)
 	retval = genl_register_family(&nfsd_nl_family);
 	if (retval)
 		goto out_free_all;
+	nfsd_localio_ops_init();
 
 	return 0;
 out_free_all: